kaiax/valset/README.md

```diff
@@ -0,0 +1,134 @@
+The council(N) is decided after the block(N).vote and govParam(N) is applied, and it will be used to calculate next committee or proposer.
+
+### committee: a subset of council who participates on consensus
+A committee(N, R) is calculated based on previous block's results(council, prevHeader, stakingInfo of council) and sorted. However, the genesis committee is copied from the genesis council.
+- minimum staking amount - The council members who have less than minimum staking amount is demoted, so it cannot be a member of committee. 
+- committeesize - The committee size can be updated via "istanbul.committeeSize". It decides the size of committee.
+- committee shuffle seed - The seed is calculated using previous block's information. The copied council is shuffled with the calculated seed to get the committee.
+
+Committee selection logic is different before/after Randao Hardfork when it's proposer policy is weightedCouncil. So the condition to activate RandaoCommittee is `policy.IsDefaultSet() || (policy.IsWeightedCouncil() && !rules.IsRandao)`
+- committee shuffle seed calculation logic
+  - before Randao: the seed is calculated using prevHash. `seed = int64(binary.BigEndian.Uint64(prevHash.Bytes()[:8]))`. 
+  - after Randao: the seed is calculated using mixHash. `seed = int64(binary.BigEndian.Uint64(mixHash[:8]))`
+- council shuffle
+  - before Randao: extract (proposer, next proposer which is differnt from proposer) and shuffle it. Attach the proposers again and slice the council.
+  - after Randao: shuffle the council and slice the committee.
+
+Example of BeforeRandaoCommittee
+```
+Condition: proposerIdx = 3, nextProposerIdx = 7, committeesize = 6, council = [0,1,2,3,4,5,6,7,8,9]
+Step1. extract proposers to committee: proposers = [3,7], council = [0,1,2,4,5,6,8,9]
+Step2. shuffle the council. proposers = [3,7], council = [4,5,6,8,9,0,1,2]
+Step3. merge. council = [3,7,4,5,6,8,9,0,1,2]
+Step4. slice the council by committee size. committee = [3,7,4,5,6,8]
+```
+### proposer: a member of committee who proposes the block
+A proposer proposes the block. We call as author after the block is created.
+#### ProposerPolicy
+Proposer is selected based on the proposer policy the network chosen. Also, each proposer selection logic has been updated per HF.
+- RoundRobin: `council(N-1)[(prevAuthorIdx+round)%len(council(N-1))]`
+- Sticky: `council(N-1)[(prevAuthorIdx+round+1)%len(council(N-1))]`
+- WeightedRandom
+  - the council is splitted into qualified, demoted addresses by minimum staking amount since istanbul HF.
+  - proposer is picked from proposers array which is created at proposerupdateinterval block with staking amount and gini. However, it is deprecated since Randao HF.
+
+The selection of proposer policy is limited by consensus algorithm.
+  - clique - 0: RoundRobin [default: 0]
+  - istanbul - 0: RoundRobin, 1: Sticky, 2: WeightedRandom [default: 0]
```

Comment:
````suggestion
### Overview

```
  GenesisValset
+ AddValidator
- RemoveValidator
------------------
  Council            kaia_getCouncil
- DemotedValidator   istanbul_getDemotedValidators
------------------
  Validators         istanbul_getValidators
* Random, Round
------------------
  Committee, Proposer    kaia_getCommittee, kaia_getBlockWithConsensusInfo
```

### Council: a list of registered GC members
A `Council(N)` indicates the council for mining the block `N`.
Thus, `Council(N)` is determined after the N-1 block has been finalized.
Note that this is usually notated as N-1 (e.g., `StakingInfo(N-1)`) in other modules.

A `Council(N)` is determined based on `header(N-1).Vote` and `gov.EffectiveParamsAt(N-1)`.
Council is utilized to calculate the next committee and the proposer.


### Committee: a subset of council who participates on consensus
A `Committee(N, R)` is randomly selected as a subset of `Council(N)` for mining the block `N` at the round `R`.
The committee is chosen based on the followings: `Council(N)`, `Header(N-1)`, and `StakingInfo(N-1)`.
The size of committee is determined by the value of the governance parameter "istanbul.committeesize", which indicates the maximum committee size.

#### Qualification
A GC member is qualified to be a member of the committee only if the following condition is met:
- Minimum staking amount: Since Istanbul HF, a council member who staked less than minimum staking amount cannot be a member of the committee.

#### Selection
The selection logic involves a shuffling algorithm, which differs depending on whether it's before/after Randao hardfork and the proposer policy.
After shuffling, the council is "sliced" up to "istanbul.committeesize".
`RandaoCommittee` logic is used when `policy.IsDefaultSet() || (policy.IsWeightedCouncil() && !rules.IsRandao)`, which corresponds to Mainnet and Kairos.
Otherwise, `SimpleRandomCommittee` is used.

- SimpleRandomCommittee: the proposer and the next proposer are secured, and the rest are shuffled and sliced.
- RandaoCommittee: shuffle the council and slice.

Specifically, the committee selection works as follows:
```
Condition: proposerIdx = 3, nextProposerIdx = 7, committeesize = 6, council = [0,1,2,3,4,5,6,7,8,9]


SimpleRandomCommittee:
- Step1. secure the proposer and the next proposer: proposers = [3,7], council = [0,1,2,4,5,6,8,9]
- Step2. shuffle the council: proposers = [3,7], council = [4,5,6,8,9,0,1,2]
- Step3. concat: council = [3,7,4,5,6,8,9,0,1,2]
- Step4. slice by committeesize: committee = [3,7,4,5,6,8]

RandaoCommittee:
- Step1. shuffle the council: council = [4,7,5,6,8,3,9,0,1,2]
- Step2. slice by committeesize: committee = [4,7,5,6,8,3]
```

The shuffling algorithm requires a "seed", which is calculated from the information of previous block;
- before Randao, `Header(N-1).hash` is used. `seed = int64(binary.BigEndian.Uint64(Header(N-1).hash.Bytes()[:8]))`
- since Randao, `Header(N-1).mixHash` is used. `seed = int64(binary.BigEndian.Uint64(Header(N-1).mixHash.Bytes()[:8]))`

### Proposer: a member of the committee who proposes the block
A `Proposer(N, R)` who proposes the block is selected from the committee.
Note that proposer is called an "author" once the block is created.

#### ProposerPolicy
Proposer is selected based on the governance parameter "istanbul.proposerpolicy".
The mechanism of the same policy works differently depending on the hardfork.

Here are the list of policies: (the number in the paranthesis indicates the value of "istanbul.proposerpolicy")
- `RoundRobin` (0): every council member takes each turn. That is, `Proposer(N, R) = Council(N-1)[(ProposerIdx(N-1) + R) % len(Council(N-1))]`.
- `Sticky` (1): every council member takes each turn. That is, `Proposer(N, R) = Council(N-1)[(ProposerIdx(N-1) + R + 1) % len(council(N-1))]`.
- `WeightedRandom` (2)
  - Since Randao HF, the proposer is randomly selected among the committee. That is, `Proposer(N, R) = Committee(N, R)[R]` .
  - Before Randao HF, the proposer is randomly selected where the probability is proportional to the gini-coeff-applied staking amount. That is, `Proposer(N, R) = proposers[(N + R) % len(proposers)]` where `proposers` is refreshed at every "reward.stakingupdateinterval".

The Mainnet and Kairos uses `WeightedRandom` policy.

The selection of proposer policy is limited by the consensus algorithm:

- clique: RoundRobin [default: RoundRobin]
- istanbul - RoundRobin, Sticky, WeightedRandom [default: WeightedRandom]
````

Reply:
@ian0371 Thanks, it helps a lot to revise the README.md. Btw, there's few things that I have a question.
* A `Council(N)` indicates the council for mining the block `N`. -> I know the getCouncil api returns like this description, but according to the valsetSnapshot(cliqueSnapshot,istanbulSnapshot) implementation, `snap(N-1).valSet` is correct because snapshot is stored after the block is processed. Do you considered it also? If so, we can change to the definition of Council as you described. https://github.com/kaiachain/kaia/blob/8f043730bf82ae3e1211cbf37988c2e20ee35b76/consensus/istanbul/backend/api.go#L235-L242
* About the `sticky`, I'm little bit confused because I thought `Sticky` is a variation of round-robin. Isn't it a round-robin?

Reply:
> A Council(N) indicates the council for mining the block N.

You're right, and we can define either way. I just thought it like the API return.
Btw, both are used in the existing use cases. These actually refer to the same "timing":
- `StakingInfo(N-1)`: finalized at N-1. Staking info required for mining N.
- `EffectiveParams(N)`: finalized at N-1. Parameters applied when mining N.

> sticky

The only difference is this line:

RoundRobin
https://github.com/kaiachain/kaia/blob/1fe135d9e12ab4ec8057aa5e71b6423353eb7834/consensus/istanbul/validator/default.go#L309
Sticky
https://github.com/kaiachain/kaia/blob/1fe135d9e12ab4ec8057aa5e71b6423353eb7834/consensus/istanbul/validator/default.go#L320

`calcSeed` is calculated based on `proposer offset` and `round`.
RoundRobin adds 1 to calcSeed, but Sticky does not.
So I thought the result of `calcSeed` is always the same unless the proposer changes.
I haven't tested it though.

I may have been wrong in case of round change; the proposer sticks to the last proposer. So it should be like below: (please confirm)

```
proposers=[0,1,2,3]
block       round           proposer
1           0               0
2           0               0
3           0               0
4           0               0
5           0               0
6           0               0
7           0               0
8           0               0
9           0               0
10          0               0
11          0               0        << round change
11          1               1        << round change
11          2               2
12          0               2
13          0               2
14          0               2
```

Reply:
I spawned 4 nodes and the simulation seems correct.

```
> kaia.getBlockWithConsensusInfo(1).proposer
"0x3c44cdddb6a900fa2b585dd299e03d12fa4293bc"
> kaia.getBlockWithConsensusInfo(2).proposer
"0x3c44cdddb6a900fa2b585dd299e03d12fa4293bc"
> kaia.getBlockWithConsensusInfo(3).proposer
"0x3c44cdddb6a900fa2b585dd299e03d12fa4293bc"
> kaia.getBlockWithConsensusInfo(4).proposer
"0x3c44cdddb6a900fa2b585dd299e03d12fa4293bc"
> kaia.getBlockWithConsensusInfo(5).proposer
"0x3c44cdddb6a900fa2b585dd299e03d12fa4293bc"
// round change
> kaia.getBlockWithConsensusInfo(56).proposer
"0x70997970c51812dc3a010c7d01b50e0d17dc79c8"
> kaia.getBlockWithConsensusInfo(55).proposer
"0x70997970c51812dc3a010c7d01b50e0d17dc79c8"
> kaia.getBlockWithConsensusInfo(56).proposer
"0x70997970c51812dc3a010c7d01b50e0d17dc79c8"
> kaia.getBlockWithConsensusInfo(57).proposer
"0x70997970c51812dc3a010c7d01b50e0d17dc79c8"
> kaia.getBlockWithConsensusInfo(58).proposer
"0x70997970c51812dc3a010c7d01b50e0d17dc79c8"
> kaia.getBlockWithConsensusInfo(59).proposer
"0x70997970c51812dc3a010c7d01b50e0d17dc79c8"
```

Reply:
I think "A Council(N) indicates the council for mining the block N" is still correct and natural.

Others are the same.
GetStakingInfo(num) returns the staking info required to make the block at `num`
EffectiveParams(num) same.

The confusing part is that istanbul.snapshot(num) affects from num+1. This seems to be the one and only exception in the kaiax modules..?

Reply:
#### 1. block number parameters to calculate getter
As we discussed in offline, the block number parameters will be as below. Updated at README.md.
```
Several dependencies are required to calculate the committee or proposer of block N:
- Header(N-1) // e.g. mixHash, blockHash
- Council(N)
- StakingInfo(N) // will read StateAt(num-1)
- EffectiveParamSet(N) // governance parameter value
```
#### 2. sticky proposer policy
It is correct that the proposer sticks to the last proposer. Updated at readme.md and added `TestPickRoundRobinProposer` unit test at proposers_test.go.